readLine: Unicode support for Windows console

ktamp opened the pull request:
When input is read from the Windows console, input encoding is UTF16. This is translated internally to UTF8.

ktamp committed:
readLine: Unicode support for Windows console

When input is read from the Windows console, input encoding is UTF16. This is translated internally to UTF8.

alaviss commented on lib/system/io.nim:
```diff
@@ -14,6 +14,11 @@ include inclrtl
 import formatfloat
 
+when defined(windows) and not defined(useWinAnsi):
+  import os
+  import system/widestrs
+  import terminal
```
This is bad, as all modules have `system/io` implicitly imported. This will cause recursive imports, which we don't support (but we should) atm.

ktamp replied:
We can work around **os** by calling **GetLastError**.
We can also work around **terminal** by accessing **File** struct contents (although this may not be pretty).
But working around **widestrs** is another matter...

This is what makes code work across compilers.
GCC has this functionality built-in, but what about other compilers?

alaviss replied:
IIRC console have special HANDLE values that we can just compare to instead.

As for widestrs, you might not even have to import it.

alaviss replied:
https://github.com/nim-lang/Nim/blob/devel/lib/pure/terminal.nim#L669-L672

`isatty()` is simple enough to just duplicate into `io` for this.

ktamp committed:
readLine: Remove recursive imports

ktamp committed:
readLine: Fix issues with --gc:arc

**--gc:arc** defines **nimv2**. This changes the definition of **WideCStringObj**.
Also an empty string should be returned in case of EOF.
